clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -879,17 +884,12 @@ private List<ProducerBatch> drainBatchesForOneNode(Metadata metadata, Node node,
             // Only proceed if the partition has no in-flight batches.
             if (isMuted(tp))
                 continue;
-            Metadata.LeaderAndEpoch leaderAndEpoch = metadata.currentLeader(tp);
-            // Although a small chance, but skip this partition if leader has changed since the partition -> node assignment obtained from outside the loop.
```

Comment:
I guess this race is no longer possible since we are using the snapshot, which is immutable?

Reply:
Correct, since its immutable, this is not possible.